Transfer schemas to remote node

clumens opened the pull request:
I have tested to verify that schemas do get transferred and added to the known schemas list, but I have not yet verified that these schemas are used to actually do the validation.  That's the next step, but in the meantime we can at least talk about what I've done up to this point.

kgaillot commented on lib/common/xml.c:
```diff
@@ -2647,11 +2647,19 @@ pcmk__xml_artefact_path(enum pcmk__xml_artefact_ns ns, const char *filespec)
         case pcmk__xml_artefact_ns_legacy_rng:
         case pcmk__xml_artefact_ns_base_rng:
-            ret = crm_strdup_printf("%s/%s.rng", base, filespec);
+            if (g_str_has_suffix(filespec, ".rng")) {
```
Huh, I wonder why I created `pcmk__ends_with()`. Maybe I didn't realize this existed, or maybe this doesn't handle NULLs.

clumens replied:
Dunno, but switched to using that.  We've got this stuff, might as well use it.

kgaillot commented on lib/common/schemas.c:
```diff
@@ -1335,3 +1335,111 @@ pcmk__schemas_later_than(const char *name)
+        return;
+    }
+
+    contents = pcmk__schema_contents(name);
```
I don't know if this would be easier, but `pcmk__schema_contents()` could return an xmlNode* to avoid the `string2xml()` call. `filename2xml()` would probably work.

clumens replied:
Whatever it returns makes something awkward.

If we return a string, we have to use `string2xml` in `external_refs_in_schema` to do the xpath query.  If we return an xmlNodePtr, we have to convert it to a string for `xmlNewCDataBlock`.

kgaillot commented on lib/common/schemas.c:
```diff
@@ -1335,3 +1335,111 @@ pcmk__schemas_later_than(const char *name)
+
+    if (g_str_has_suffix(name, ".rng")) {
+        crm_xml_add(file_node, PCMK__XA_PATH, name);
+        *already_included = g_list_prepend(*already_included, strdup(name));
```
Worth doing the strdup() separately to assert non-null

clumens replied:
Fixed, added an assert in the else case as well, and switched to using pcmk__ends_with.

kgaillot reviewed (commented):
I didn't get through the whole thing but it's looking good so far

clumens commented on .gitignore:
```diff
@@ -143,6 +143,7 @@ configure
 /include/config.h
 /include/config.h.in
 /include/crm_config.h
+/lib/common/tests/schemas/pcmk__schema_contents_test.c
```
Of course.  I don't know why I went with the substitution method first.  I'm sure it made sense at the time.

clumens commented on lib/common/schemas.c:
```diff
@@ -1275,3 +1275,30 @@ cli_config_update(xmlNode **xml, int *best_version, gboolean to_logs)
+    int rc = pcmk_rc_ok;
+    char *contents = NULL;
+    char *path = pcmk__xml_artefact_path(pcmk__xml_artefact_ns_legacy_rng,
+                                         name);
```
I'd kind of disregarded this approach at first because I thought `pcmk__xml_artefact_path` was used for constructing paths to write to in addition to constructing search paths.  That turns out to not be the case, so I've added PCMK_remote_schema_directory (which needs another patch to edit the sysconfig file, I'll do that in a sec).  This also allows me to drop the patch that keeps track of where schema files come from.

clumens commented:
Updated to fix a couple covscan problems.

clumens commented on lib/common/xml.c:
```diff
@@ -2639,33 +2640,54 @@ pcmk__xml_artefact_root(enum pcmk__xml_artefact_ns ns)
+    ret = find_artefact(ns, base, filespec);
     free(base);
 
+    if (stat(ret, &sb) != 0 || !S_ISREG(sb.st_mode)) {
```
This could potentially be a little function somewhere - `pcmk__file_exists`, for instance.  crm_shadow does something similar.

kgaillot replied:
This actually is whether it's a directory, probably not worth functionizing now

clumens commented on daemons/execd/pacemaker-execd.c:
```diff
@@ -400,6 +404,66 @@ handle_shutdown_nack(void)
+    }
+}
+#endif
+
```
There's a couple things in here that could potentially be moved into library functions:

* Removing a directory tree
* Creating a directory if it doesn't exist
* Testing if a directory exists or not

clumens commented on daemons/execd/remoted_tls.c:
```diff
@@ -38,12 +42,190 @@ gnutls_dh_params_t dh_params;
+            write_schema_file(file_node, CRM_REMOTE_SCHEMA_DIRECTORY);
+            need_to_load = true;
+        }
+
```
Here's another function idea - a foreach function for xpath matches.

clumens replied:
Oh, this already exists - `crm_foreach_xpath_result`.  Guess I'll see if it's useful in this case.

clumens commented on lib/common/schemas.c:
```diff
@@ -1331,3 +1331,115 @@ pcmk__schemas_later_than(const char *name)
+            char *s = strdup(href);
+
+            CRM_ASSERT(s != NULL);
+            *list = g_list_append(*list, s);
```
Incidentally, I know we typically like to use g_list_prepend and then reverse.  There aren't a ton of external refs in any schema, but it's not nothing.  On the other hand, we're not exactly committed to using g_list_prepend...
```
clumens@spire ~/src/pacemaker schemas> grep -r g_list_prepend | wc -l 
91
clumens@spire ~/src/pacemaker schemas> grep -r g_list_append | wc -l
136
```

But maybe it's not worth making that problem worse.

clumens commented on lib/common/schemas.c:
```diff
@@ -1331,3 +1331,115 @@ pcmk__schemas_later_than(const char *name)
+     */
+    for (GList *iter = includes; iter != NULL; iter = iter->next) {
+        add_schema_file_to_xml(parent, iter->data, already_included);
+    }
```
I could also change this to use `g_list_foreach`, which would require changing the definition of `add_schema_file_to_xml` and adding a little struct somewhere for the user data.  I dunno if it's more readable or not, but I kind of like being in the habit of using the foreach functions.  Meh, I could go either way with this.

clumens replied:
After playing around with this, I think it's less readable to use g_list_foreach.

kgaillot commented on lib/common/tests/schemas/pcmk__schemas_later_than_test.c:
```diff
@@ -0,0 +1,86 @@
+    /* And then something way in the future.  Hopefully this is far enough
+     * out to where we don't have to change this test any time soon.
+     */
+    schemas = pcmk__schemas_later_than("pacemaker-9.0");
```
999 to be really safe :)

Alternatively, create a test schema directory with symlinks to a few schemas and use that

clumens replied:
Yeah, I could create lib/common/tests/schemas/schemas, add a couple symlinks in there, and then set PCMK__TEST_SCHEMA_DIR to that.  This could probably also all be done at build time, without putting the symlinks in git.  For instance, I could create the directory and symlink over the ten oldest schema files.  Then I'd know what to expect and we wouldn't be chasing whatever happens with the newest ones.  I think we'd still be okay from a thoroughness of testing perspective.

clumens replied:
Updated to add this.

kgaillot reviewed (commented):
Didn't get through whole thing (again) but first half looks good

clumens commented on lib/common/schemas.c:
```diff
@@ -1302,3 +1302,32 @@ pcmk__schema_contents(const char *name)
+
+    for (int lpc = xml_latest_schema_index(); lpc > 0; lpc--) {
+        if (schema_sort(ver, known_schemas[lpc].version) == -1) {
+            lst = g_list_prepend(lst, strdup(known_schemas[lpc].name));
```
Fixed.  Note that there's hundreds of spots we are calling strdup() without checking.

kgaillot replied:
Yep we're gradually fixing that as we come across them. At some point it may be worth a project just for that. Aside from a controlled exit, the main benefit is to minimize static analysis complaints

clumens commented on lib/common/tests/schemas/Makefile.am:
```diff
@@ -0,0 +1,18 @@
+include $(top_srcdir)/mk/tap.mk
+include $(top_srcdir)/mk/unittest.mk
+
+CFLAGS += -DPCMK__TEST_SCHEMA_DIR=$(top_builddir)/xml
```
I have no strong preferences, but using the quoting seems like the easiest approach.

clumens commented on include/crm_internal.h:
```diff
@@ -81,13 +81,17 @@
 #define PCMK__XA_GRAPH_WARNINGS         "graph-warnings"
 #define PCMK__XA_MODE                   "mode"
 #define PCMK__XA_NODE_START_STATE       "node_start_state"
+#define PCMK__XA_PATH                   "path"
```
I could do whatever here.  It's kind of unfortunate to have the same definition in two places, let alone three.

Additionally, I just remembered I'm using `crm_element_value(xml, "href")` somewhere, which I thought would have already been defined, but is not.  I could add a definition for that too.

kgaillot replied:
whichever you prefer

kgaillot commented on configure.ac:
```diff
@@ -838,6 +838,11 @@ AC_DEFINE_UNQUOTED([CRM_SCHEMA_DIRECTORY], ["$CRM_SCHEMA_DIRECTORY"],
+CRM_REMOTE_SCHEMA_DIRECTORY="${localstatedir}/lib/pacemaker/schemas"
+AC_DEFINE_UNQUOTED([CRM_REMOTE_SCHEMA_DIRECTORY], ["$CRM_REMOTE_SCHEMA_DIRECTORY"],
+                   [Location to store Relax-NG Schema files on remote nodes])
+AC_SUBST(CRM_REMOTE_SCHEMA_DIRECTORY)
```
ah cool

kgaillot reviewed (commented):
Finally got through it :) This is great, going to be a cool improvement

clumens commented on lib/common/schemas.c:
```diff
@@ -372,6 +372,53 @@ wrap_libxslt(bool finalize)
+    }
+
+    for (lpc = 0; lpc < max; lpc++) {
+        bool transform_expected = FALSE;
```
Yeah, this was code that was in crm_schema_init and got copied over.  Updated.

clumens commented on daemons/execd/pacemaker-execd.c:
```diff
@@ -452,6 +516,12 @@ main(int argc, char **argv, char **envp)
+    // crm_log_preinit is called because that will end up calling crm_schema_init,
+    // which will in turn read all the existing schema files including any in
+    // PCMK__REMOTE_SCHEMA_DIR That could cause us to think we already have all
+    // the files we need, when in fact we want to fetch them again every time.
```
Right, just trying to follow the local style (there's already a // comment immediately below this).  But whatever, changed.

clumens commented on lib/cib/cib_ops.c:
```diff
@@ -127,6 +127,9 @@ static const cib__operation_t cib_ops[] = {
         |cib__op_attr_transaction
     },
+    {
+        PCMK__CIB_REQUEST_SCHEMAS, cib__op_schemas, cib__op_attr_none
```
Fixed.

clumens commented on lib/common/tests/schemas/pcmk__schemas_later_than_test.c:
```diff
@@ -0,0 +1,71 @@
+
+    /* And then something way in the future.  Hopefully this is far enough
+     * out to where we don't have to change this test any time soon.
+     */
```
Changed.

clumens commented on lib/common/tests/schemas/Makefile.am:
```diff
@@ -0,0 +1,41 @@
+# tests without having to always make changes when new schemas are added.
+#
+# The "head -14" is to pull in everything through pacemaker-2.10.rng, needed so
+# we can check that versions are compares as numbers and not as strings.
```
Fixed.

clumens commented on lib/common/tests/schemas/pcmk__schema_contents_test.c:
```diff
@@ -0,0 +1,55 @@
+
+static int
+setup(void **state) {
+    setenv("PCMK_schema_directory", "schemas", 1);
```
Fixed.

clumens commented on daemons/based/based_messages.c:
```diff
@@ -479,3 +479,51 @@ cib_process_commit_transaction(const char *op, int options, const char *section,
+     */
+    if (schemas == NULL) {
+        crm_err("No schemas found after version '%s'", after_ver);
+        return -EINVAL;
```
Right, this is what the comment directly above was asking.  Removed this block.

clumens commented on lib/common/xml.c:
```diff
@@ -2610,33 +2611,54 @@ pcmk__xml_artefact_root(enum pcmk__xml_artefact_ns ns)
-char *
-pcmk__xml_artefact_path(enum pcmk__xml_artefact_ns ns, const char *filespec)
+static char *
+find_artefact(enum pcmk__xml_artefact_ns ns, const char *path, const char *filespec)
```
Updated.

clumens commented on lib/common/schemas.c:
```diff
@@ -147,7 +147,20 @@ schema_filter(const struct dirent *a)
 
 static int
-schema_sort(const struct dirent **a, const struct dirent **b)
+schema_sort(schema_version_t a_version, schema_version_t b_version)
```
Fixed.

clumens commented on daemons/execd/remoted_tls.c:
```diff
@@ -38,12 +42,169 @@ gnutls_dh_params_t dh_params;
+            continue;
+        }
+
+        rc = pcmk__write_sync(fd, (const char *) child->content);
```
Fixed.

clumens commented on lib/common/schemas.c:
```diff
@@ -1275,3 +1275,30 @@ cli_config_update(xmlNode **xml, int *best_version, gboolean to_logs)
+    rc = pcmk__file_contents(path, &contents);
+
+    if (rc != pcmk_rc_ok) {
+        crm_err("Could not read schema file %s: %s", path, pcmk_rc_str(rc));
```
Changed, and test cases updated.

clumens commented on lib/common/schemas.c:
```diff
@@ -372,6 +372,53 @@ wrap_libxslt(bool finalize)
+
+            if (version_from_filename(namelist[lpc+1]->d_name, &next_version)
+                    && (version.v[0] < next_version.v[0])) {
+                transform_expected = TRUE;
```
Ah, right, that came up in unit testing too.  I'll add something here.  For the moment, I've also replaced that `TRUE` with `true`.

clumens commented on daemons/execd/remoted_tls.c:
```diff
@@ -97,6 +258,11 @@ lrmd_remote_client_msg(gpointer data)
+    if (try_getting_schema_files) {
+        try_getting_schema_files = false;
+        request_cib_schema_files();
+    }
```
Removed the guard.

kgaillot commented on daemons/execd/remoted_tls.c:
```diff
@@ -38,12 +42,169 @@ gnutls_dh_params_t dh_params;
+    /* If the referenced path includes a '/', build up the expected directory
+     * structure under PCMK__REMOTE_SCHEMA_DIR.
+     */
+    if (strchr(file, '/') != NULL) {
```
Yeah, we don't actually need to copy that one over. It doesn't hurt but it might drag in a bunch of api/ unnecessarily. I've been wanting to move the CIB files to an xml/cib/ subdirectory anyway (first part of https://projects.clusterlabs.org/T8 ) so maybe this is a good time to do it.

clumens replied:
Right, we don't need to copy that one, and I think that is the only href that uses a subdirectory.  Nothing should pull in api/ because nothing in the cib-related files references anything in it.  I'm fine with removing all this subdirectory crud as long as we can remember the rule that schema files can't use subdirectories.

clumens replied:
Removed.

clumens commented:
Updated with a lot of changes:

* Addressed review comments
* Moved all the new remoted schema stuff into its own file
* Got rid of pcmk__schema_contents
* Added more files to the test schema directory
* Added unit tests for pcmk__build_schema_xml_node
* Transfer XSLT files as well
* Changed where remoted asks for schema files to avoid fork bombing the system
* Changed where cleanup happens to address review

kgaillot commented on daemons/based/based_messages.c:
```diff
@@ -478,3 +478,46 @@ cib_process_commit_transaction(const char *op, int options, const char *section,
+
+    data = get_message_xml(req, F_CIB_CALLDATA);
+    if (data == NULL) {
+        crm_err("No data specified in request");
```
These can be warnings since the CIB manager itself is unaffected

kgaillot commented on daemons/execd/remoted_tls.c:
```diff
@@ -131,6 +133,12 @@ lrmd_remote_client_msg(gpointer data)
+        if (pcmk__str_eq(op, CRM_OP_REGISTER, pcmk__str_none)) {
+            remoted_request_cib_schema_files();
+        }
+
```
I think it makes sense to do this in process_lrmd_signon() after ipc_proxy_add_provider()

kgaillot commented on daemons/execd/remoted_schemas.c:
```diff
@@ -0,0 +1,229 @@
+
+        default:
+            /* parent */
+            mainloop_child_add(pid, 0, "schema-fetch", NULL, NULL);
```
A timeout even if it's longish would help if there's some sort of recovery loop for the remote connection and the children hang for whatever reason. For completeness it might be worth tracking whether there's a child active and kill it (and clean the directory) before forking a new one.

clumens replied:
What kind of thing are you envisioning here?  I've not had any luck getting the client set up in any way where it can actually be killed when the timeout expires - I don't think you can kill a sleeping process (or at least, I haven't been able to), and I think if the child is waiting to hear back from the server, it'll be stuck waiting on IO which I also don't think is killable.

kgaillot replied:
If you pass a timeout here, the library will automatically handle the killing in that case. Even if the process doesn't immediately go away, it prevents it from doing anything further if it wakes up, and allows it to be reaped if the parent dies without waiting for it.

For the case where a previous child is still active when this gets called again, `mainloop_child_kill()` should work the same.

clumens replied:
That's not what I am seeing, unless I'm doing something wrong.  For instance, with the following patch:
```
diff --git a/daemons/execd/remoted_schemas.c b/daemons/execd/remoted_schemas.c
index 7be0d30fc..28ead991b 100644
--- a/daemons/execd/remoted_schemas.c
+++ b/daemons/execd/remoted_schemas.c
@@ -132,6 +132,8 @@ get_schema_files(void)
     cib_t *cib = NULL;
     xmlNode *reply;
 
+    sleep(10);
+
     cib = cib_new();
     if (cib == NULL) {
         _exit(ENOTCONN);
@@ -243,7 +245,7 @@ remoted_request_cib_schema_files(void)
 
         default:
             /* parent */
-            mainloop_child_add(pid, 0, "schema-fetch", NULL, get_schema_files_complete);
+            mainloop_child_add(pid, 1 * 1000, "schema-fetch", NULL, get_schema_files_complete);
             break;
     }
 }
```

I see the following in the logs on the remote node:
```
Nov 16 13:56:12.660 rhel9-cluster-3 pacemaker-remoted   [773845] (child_waitpid@mainloop.c:1120)        trace: Child process 773850 (schema-fetch) still active
Nov 16 13:56:12.991 rhel9-cluster-3 pacemaker-remoted   [773845] (child_waitpid@mainloop.c:1120)        trace: Child process 773850 (schema-fetch) still active
Nov 16 13:56:13.097 rhel9-cluster-3 pacemaker-remoted   [773845] (child_waitpid@mainloop.c:1120)        trace: Child process 773850 (schema-fetch) still active
...
Nov 16 13:56:15.642 rhel9-cluster-3 pacemaker-remoted   [773845] (child_waitpid@mainloop.c:1120)        trace: Child process 773850 (schema-fetch) still active
Nov 16 13:56:22.726 rhel9-cluster-3 pacemaker-remoted   [773845] (child_waitpid@mainloop.c:1141)        trace: Child process 773850 (schema-fetch) exited with status 0
```

And the schema files do get downloaded.

clumens replied:
Yeah, kill is returning ESRCH when it is called on the group:
```
Nov 16 15:00:31.561 rhel9-cluster-3 pacemaker-remoted   [6167] (child_timeout_callback@mainloop.c:1093)         info: CHILD_TIMEOUT_CALLBACK CALLED FOR 6169 (schema-fetch)
Nov 16 15:00:31.561 rhel9-cluster-3 pacemaker-remoted   [6167] (child_kill_helper@mainloop.c:1067)      info: CHILD_KILL_HELPER CALLED FOR schema-fetch
Nov 16 15:00:31.561 rhel9-cluster-3 pacemaker-remoted   [6167] (child_kill_helper@mainloop.c:1073)      info: Kill pid 6169's group
Nov 16 15:00:31.561 rhel9-cluster-3 pacemaker-remoted   [6167] (child_kill_helper@mainloop.c:1078)      info: RETURNING -3
```

kgaillot replied:
Ahhh, the group kill only works if the child calls `setpgid(0, 0)`. Either that or adding the child with `mainloop_leave_pid_group` should work. The latter is probably what we want since we don't expect this child to fork any  children of its own.

clumens replied:
I think I've got this figured out, then.

kgaillot commented on daemons/execd/remoted_schemas.c:
```diff
@@ -0,0 +1,229 @@
+    crm_schema_cleanup();
+    crm_schema_init();
+
+    rc = cib->cmds->fetch_schemas(cib, &reply, xml_latest_schema(), cib_sync_call);
```
What happens if the cluster node doesn't support the new command? Ideally we'd check before attempting to fetch, but if older daemons will reply with an error, that would be fine too.

kgaillot replied:
We do exchange `LRMD_PROTOCOL_VERSION` in the handshake. We could (mis)use that to determine CIB protocol support.

Background: `LRMD_PROTOCOL_VERSION` was initially intended for "do we allow this remote connection at all". We bumped it by mistake, so we added `LRMD_MIN_PROTOCOL_VERSION` to work around that. That could be useful though -- we could use `LRMD_MIN_PROTOCOL_VERSION` for connection support, and `LRMD_PROTOCOL_VERSION` for specific feature support. Or, maybe an even better approach would be to use the protocol major version for connection compatibility and the minor version for feature support.

clumens replied:
Added a commit on top to take care of this.  I'm going with the latter approach, where the major version is for connection compatibility and the minor version is for features.

kgaillot commented on daemons/execd/remoted_schemas.c:
```diff
@@ -0,0 +1,229 @@
+        /* Don't just crm_schema_init here because that will load the base
+         * schemas again too.  Instead just load the things we fetched.
+         */
+        pcmk__load_schemas_from_dir(PCMK__REMOTE_SCHEMA_DIR);
```
This will only affect the child process. We probably need a callback to do this in the parent when the child completes. (The cleanup and re-init probably should be done in the parent before forking.)

kgaillot commented on daemons/execd/remoted_schemas.c:
```diff
@@ -0,0 +1,229 @@
+        if (stream == NULL) {
+            crm_warn("Could not write schema file %s: %s", path, strerror(errno));
+        } else {
+            rc = fprintf(stream, "%s", child->content);
```
When you test this, do an md5sum on the downloaded schemas to see if they're identical. It's possible there may be whitespace differences or whatnot but that's worth checking.

clumens replied:
Are you looking for comparing the md5sum of child->content against the md5sum that the server said the content should be?  If that's the case, I'll need to add it as an attribute to the <file> node.  I don't think we add that automatically anywhere at the moment.

kgaillot replied:
No, not in the code, just as a manual comparison for testing purposes

clumens replied:
These all look good.

kgaillot commented on daemons/execd/remoted_schemas.c:
```diff
@@ -0,0 +1,229 @@
+                rc = errno;
+                crm_err("Could not create directory for schemas: %s",
+                        pcmk_rc_str(rc));
+            }
```
presumably you want to start over if this succeeds? right now it would just return

clumens replied:
I'm not sure I quite follow - it doesn't help that I'm using the same message twice and github doesn't display a lot of context.  There's several cases to consider here:

* The directory doesn't exist, and mkdir succeeds - we don't need to do anything else here.
* The directory doesn't exist, and mkdir fails - I don't think there's anything we could possibly do.  I think everything else involving the directory will fail properly.  It would be better if it didn't even try, though.
* stat returned some other error code - ???.  This could be that a file exists with the same name already, or we don't have permission to check, or all sorts of things.  I'm currently catching the case that something with the same name already exists by simply not doing anything.  I don't know what to do with all the other possibilities.

kgaillot replied:
Ah, what I missed is that if we have to create the directory, it's already clean :)

kgaillot reviewed (commented):
Looking good

clumens commented on lib/common/schemas.c:
```diff
@@ -1275,3 +1281,63 @@ cli_config_update(xmlNode **xml, int *best_version, gboolean to_logs)
+        char *s = NULL;
+
+        if (schema_cmp(ver, known_schemas[lpc].version) != -1) {
+            continue;
```
I think we actually have to traverse the whole list of schemas everywhere now.  Everything had a built-in assumption that because we were only adding schemas once from one directory, the list we called add_schema on would already be sorted (via scandir and schema_cmp_directory).  add_schema doesn't do anything to preserve order, so adding a second set of schemas is just going to cause them to get added to the end of the list.

If that's going to be a problem, I'll need to change add_schema to keep the list sorted.

kgaillot replied:
Yeah, I suspect we'll have unexpected problems if we don't keep it sorted

clumens replied:
The thing that's tripping up sorting at the moment is that each schema record holds the index of the schema record that it can be upgraded to.  We can replace the index with a pointer to get around that.  But I'm also wondering if there's any instances (besides "pacemaker-next" and "none") where we care.  At the moment, every schema simply upgrades to the next one in the list.  If we can assume that will always be the case, I can just get rid of this index/pointer entirely.

kgaillot replied:
That makes sense

clumens replied:
Added a commit on top to deal with this.  #3247 will need to be merged first, however.

clumens commented:
Pushing an update to clear up the file conflicts and deal with most review comments.  I still need to address the timeout and give it one more test.

kgaillot commented on lib/common/tests/schemas/Makefile.am:
```diff
@@ -10,7 +10,47 @@
+
+.PHONY: clean-local
+clean-local:
+	-rm -rf schemas
```
gitignore this too

kgaillot commented on lib/common/schemas.c:
```diff
@@ -1351,6 +1351,148 @@ pcmk__schema_files_later_than(const char *name)
+     */
+    if (schema_node->children != NULL) {
+        xmlAddChild(parent, schema_node);
+    }
```
Don't we need to free schema_node otherwise?

clumens replied:
Yep, added.

kgaillot commented on daemons/execd/remoted_schemas.c:
```diff
@@ -0,0 +1,248 @@
+            continue;
+        }
+
+        stream = fopen(path, "w+");
```
FYI, I believe we set a umask turning off group write and other read/write, which should be fine for anything that needs schemas (even ACL-limited users will be in the HA group).

kgaillot commented on daemons/execd/remoted_schemas.c:
```diff
@@ -0,0 +1,248 @@
+        /* Don't just crm_schema_init here because that will load the base
+         * schemas again too.  Instead just load the things we fetched.
+         */
+        pcmk__load_schemas_from_dir(PCMK__REMOTE_SCHEMA_DIR);
```
Shouldn't this check the environment variable first?

clumens replied:
Which environment variable do you mean?  I didn't introduce a new one for where extra schemas get downloaded - it's just a configure option.

clumens replied:
Oh, you mean the /etc/sysconfig/pacemaker bit.  Right.

kgaillot commented on lib/common/schemas.c:
```diff
@@ -465,6 +465,7 @@ crm_schema_init(void)
     wrap_libxslt(false);
 
     pcmk__load_schemas_from_dir(base);
+    pcmk__load_schemas_from_dir(PCMK__REMOTE_SCHEMA_DIR);
```
Environment variable?

kgaillot commented on etc/sysconfig/pacemaker.in:
```diff
@@ -339,6 +339,13 @@
+# RNG schemas and XSL transforms fetched from the cluster.
+#
+# Default: PCMK_remote_schema_directory="@PCMK__REMOTE_SCHEMA_DIR@"
+
```
Also add to new doc/sphinx/Pacemaker_Explained/local-options.rst

kgaillot commented on include/crm/lrmd.h:
```diff
@@ -46,6 +51,8 @@ typedef struct lrmd_key_value_s {
  */
 #define LRMD_MIN_PROTOCOL_VERSION "1.0"
```
We can stop using and deprecate this constant, and instead compare major versions

kgaillot replied:
It's coincidental that MIN is a .0 version, but it's a convenient coincidence. If we make it a macro, we don't have worry about ever having to update it when the main version changes.

kgaillot reviewed (commented):
Awesome :)

kgaillot commented on doc/sphinx/Pacemaker_Explained/local-options.rst:
```diff
@@ -478,6 +478,18 @@ environment variables when Pacemaker daemons start up.
+     - |PCMK__REMOTE_SCHEMA_DIR|
+     - *Advanced Use Only:* Specify an alternate location on older Pacemaker
+       Remote nodes for storing newer RNG schemas and XSL transforms fetched
+       from the cluster.
```
I'd drop "older", it might be confusing

kgaillot commented on daemons/execd/remoted_schemas.c:
```diff
@@ -0,0 +1,295 @@
+        if (killed == FALSE) {
+            crm_warn("Unable to kill pre-existing schema-fetch process");
+            return;
+        }
```
Did you mean `killed = mainloop_...`?

clumens replied:
Yes.

kgaillot commented on daemons/execd/remoted_schemas.c:
```diff
@@ -0,0 +1,295 @@
+            schema_fetch_pid = pid;
+            mainloop_child_add_with_flags(pid, 5 * 60 * 1000, "schema-fetch", NULL,
+                                          mainloop_leave_pid_group,
+                                          get_schema_files_complete);
```
Probably worth logging that we have requested any new schema files, and then log when it completes. That way if anyone has any problems in the time between, they can figure out why (or if there were no new schemas, they can rule that out).

kgaillot reviewed (commented):
Looking good

clumens commented on lib/common/xml.c:
```diff
@@ -2610,33 +2611,54 @@ pcmk__xml_artefact_root(enum pcmk__xml_artefact_ns ns)
+        const char *extra = pcmk__env_option(PCMK__ENV_REMOTE_SCHEMA_DIR);
+        if (extra == NULL) {
+            extra = PCMK__REMOTE_SCHEMA_DIR;
+        }
```
What would be nice is if pcmk__env_option took a fallback value, since that's really what I'm doing here.

clumens commented:
I'm ready.
